.github/generate-doc.py

```diff
@@ -12,6 +14,137 @@ class Colors:
+
+    with open(file_path, 'w', encoding='utf-8') as file:
+        file.write(modified_content)
+    return modified_content
+
+
+def extract_subsection(file_path, target_subsection_title):
+    html_subsection_pattern = re.compile(
+        r'<h2>(.*?)<\/h2>', re.IGNORECASE | re.DOTALL)
+    with open(file_path, 'r', encoding='utf-8') as file:
+        markdown_text = file.read()
+    html_subsection_matches = html_subsection_pattern.finditer(markdown_text)
+
+    # Iterate over matches to find the target subsection
+    for match in html_subsection_matches:
+        current_subsection_title = match.group(1).strip()
+
+        if current_subsection_title.lower() == target_subsection_title.lower():
+
+            subsection_start = match.start()
+            next_subsection_match = next(html_subsection_matches, None)
+            subsection_end = next_subsection_match.start(
+            ) if next_subsection_match else len(markdown_text)
+
+            subsection_content = markdown_text[match.end(
+            ):subsection_end].strip()
+
+            subsubsection_matches = html_subsection_pattern.finditer(
+                subsection_content)
+            subsubsections = [{"title": submatch.group(1).strip(
+            ), "content": ''} for submatch in subsubsection_matches]
+
+            return {
+                "title": current_subsection_title,
+                "content": subsection_content,
+                "subsubsections": subsubsections
+            }
+
+    return None
+
+
+def update_doc(readme_path, reference_path):
+    placeholder = [
+        '\n',
+        '## References',
+        '\n',
+        'Empty substring created',
+        '\n',
+        '### Subsubsection',
+        '\n',
+        'Placeholder']
+    updated = False
+
+    # Create the dir in case documentation is not created yet
+    directory_path = os.path.dirname(readme_path)
+    os.makedirs(directory_path, exist_ok=True)
+
+    # Create Readme if it does not exist yet
+    ci_name = os.path.basename(directory_path)
+    file_exist = os.path.exists(readme_path)
+    if not file_exist:
+        with open(readme_path, 'w') as file:
+            file.write(f"# Documentation for {ci_name}")
+            file.writelines(placeholder)
+
+    with open(readme_path, 'r', encoding='utf-8') as file1:
+        readme_content = file1.read()
+
+    target_subsection_title = 'References'
+
+    # add subsection if it does not exist
+    if not readme_content.__contains__(f"## {target_subsection_title}") and not readme_content.__contains__(f"<h2>{target_subsection_title}</h2>"):
+        with open(readme_path, 'a', encoding='utf-8') as file_readme:
+            file_readme.writelines(placeholder)
+    format_markdown_file(readme_path)
+    readme_extraction_result = extract_subsection(
+        readme_path, target_subsection_title)
+
+    reference_extraction_result = extract_subsection(
+        reference_path, target_subsection_title)
+
+    if file_exist:
```

Comment:
Is this check necessary? You are creating the file above if it does not exist, right?